src/herbpy/herbrobot.py

```diff
@@ -242,35 +319,187 @@ def CloneBindings(self, parent):
+                active_controllers.append(
+                    RewdOrTrajectoryController(self, '',
+                                               'right_trajectory_controller',
+                                               self.right_arm.GetJointNames()))
+
+            if 'left_trajectory_controller' in controllers_manip:
+                active_controllers.append(
+                    RewdOrTrajectoryController(self, '',
+                                               'left_trajectory_controller',
+                                               self.left_arm.GetJointNames()))
+
+        if needs_base:
+            if (hasattr(self, 'base') and hasattr(self.base, 'controller') and
+                    self.base.controller is not None):
+                active_controllers.append(self.base.controller)
+            else:
+                logger.warning(
+                    'Trajectory includes the base, but no base controller is'
+                    ' available. Is self.base.controller set?')
+
+        for controller in active_controllers:
+            controller.SetPath(traj)
+
+        if defer is True:
+            import time
+            import trollius
+
+            @trollius.coroutine
+            def do_poll():
+                time_stop = time.time() + (timeout if timeout else numpy.inf)
+
+                while time.time() <= time_stop:
+                    is_done = all(controller.IsDone()
+                                  for controller in active_controllers)
+                    if is_done:
+                        raise trollius.Return(traj)
 
-        active_manipulators = self.GetTrajectoryManipulators(traj)
+                    yield trollius.From(trollius.sleep(period))
 
-        for manipulator in active_manipulators:
-            manipulator.ClearTrajectoryStatus()
+                raise trollius.Return(None)
 
-        value = super(HERBRobot, self).ExecuteTrajectory(traj, *args, **kwargs)
+            return trollius.async(do_poll())
+        elif defer is False:
+            prpy.util.WaitForControllers(active_controllers, timeout=timeout)
+            return traj
+        else:
+            raise ValueError('Received unexpected value "{:s}" for defer.'
+                             .format(str(defer)))
+
+    def ExecuteTrajectory(self, traj, *args, **kwargs):
+        # from prpy.exceptions import TrajectoryAborted
 
-        for manipulator in active_manipulators:
-            status = manipulator.GetTrajectoryStatus()
-            if status == 'aborted':
-                raise TrajectoryAborted()
+        value = self._ExecuteTrajectory(traj, *args, **kwargs)
+
+        # TODO meaningful to do this check here?
+        # if so can be done on value future
+        #
+        # for manipulator in active_manipulators:
+        #     status = manipulator.GetTrajectoryStatus()
+        #     if status == 'aborted':
+        #         raise TrajectoryAborted()
```

Comment:
We should replace this check with code that tests whether any of the `FollowJointTrajectory` actions failed, i..e. by inspecting the `Future` returned by the controller client.